# Data Quality Analysis Using Hypothesis Testing Techniques


### Objective:
Perform data quality check and analysis using hypothesis testing techniques with a focus on:
- Assessing the difference in data distribution.
- Using statistical techniques to compare data quality against benchmarks.


### 1. Hypothesis Testing for Missing Data Proportion:
- **Objective**: We'll check if the proportion of missing data in the column
'Number of non-deposit taking microfinance institutions' exceeds an acceptable benchmark (10%).
- **Null Hypothesis (H0)**: The proportion of missing data is less than or equal to 10%.
- **Alternative Hypothesis (H1)**: The proportion of missing data exceeds 10%.


### 2. Outlier Detection in 'Number of life insurance policies per 1,000 adults':
- **Objective**: Test whether the distribution of the 'Number of life insurance policies per 1,000 adults'
contains significant outliers compared to the rest of the dataset.
- **Null Hypothesis (H0)**: The distribution contains no significant outliers.
- **Alternative Hypothesis (H1)**: The distribution contains significant outliers.







In [1]:
# Importing necessary libraries
import pandas as pd
import numpy as np
from scipy import stats

C:\Users\Anshu Pandey\AppData\Roaming\Python\Python310\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


In [2]:
# Load the cleaned India FAS data
india_data_cleaned = pd.read_csv('datasets-1/india_fas_cleaned.csv')

In [4]:
india_data_cleaned

Economy  Year ISO-3 code  Institutions of commercial banks  \
0    India  2004        IND                             291.0   
1    India  2005        IND                             289.0   
2    India  2006        IND                             222.0   
3    India  2007        IND                             182.0   
4    India  2008        IND                             176.0   
5    India  2009        IND                             172.0   
6    India  2010        IND                             169.0   
7    India  2011        IND                             171.0   
8    India  2012        IND                             173.0   
9    India  2013        IND                             157.0   
10   India  2014        IND                             151.0   
11   India  2015        IND                             152.0   
12   India  2016        IND                             153.0   
13   India  2017        IND                             159.0   
14   India  2018        IND                             161.0   
15   India  2019        IND                             161.0   
16   India  2020        IND                             151.0   
17   India  2021        IND                             139.0   
18   India  2022        IND                             141.0   
19   India  2023        IND                             140.0   

    Number of credit unions and credit cooperatives  \
0                                          108848.0   
1                                          111797.0   
2                                          109354.0   
3                                           96155.0   
4                                           97840.0   
5                                           98475.0   
6                                           97441.0   
7                                           96177.0   
8                                           95186.0   
9                                           96230.0   
10                                          95721.0   
11                                          95745.0   
12                                          95984.0   
13                                          98174.0   
14                                          97799.0   
15                                          98550.0   
16                                          98047.0   
17                                         105093.0   
18                                         106783.0   
19                                         106773.0   

    Number of other deposit takers  \
0                            604.0   
1                            507.0   
2                            428.0   
3                            401.0   
4                            364.0   
5                            336.0   
6                            308.0   
7                            297.0   
8                            271.0   
9                            254.0   
10                           241.0   
11                           220.0   
12                           202.0   
13                           172.0   
14                           110.0   
15                            88.0   
16                            69.0   
17                            58.0   
18                            49.0   
19                            39.0   

    Number of non-deposit taking microfinance institutions  \
0                                                 NaN        
1                                                 NaN        
2                                                 NaN        
3                                                 NaN        
4                                                 NaN        
5                                                 NaN        
6                                                 NaN        
7                                                41.0        
8                                                41.0        
9                                                46.0        


### 1. Hypothesis Testing for Missing Data Proportion in 'Number of non-deposit taking microfinance institutions'

- **Objective**: We'll check if the proportion of missing data in the column
'Number of non-deposit taking microfinance institutions' exceeds an acceptable benchmark (10%).
- **Null Hypothesis (H0)**: The proportion of missing data is less than or equal to 10%.
- **Alternative Hypothesis (H1)**: The proportion of missing data exceeds 10%.

In [5]:
# Set benchmark for acceptable missing data proportion (10%)
benchmark_proportion = 0.10

# Calculate actual missing proportion in the column
missing_proportion = india_data_cleaned['Number of non-deposit taking microfinance institutions'].isna().mean()
print(missing_proportion)

0.35


In [6]:
from statsmodels.stats.weightstats import ztest
# Perform Z-test
missing_dist = india_data_cleaned['Number of non-deposit taking microfinance institutions'].isna().astype(int)
z_score, p_value = ztest(missing_dist.values, value=benchmark_proportion)
print(f"P-value: {p_value}")

P-value: 0.02233137323552469


In [9]:
z_score

2.2846838336584385

In [7]:
# Perform one-sample z-test for proportions
n = len(missing_dist)
p_hat = missing_proportion # sample mean
p_0 = benchmark_proportion # populaiton mean
s_std = np.std(missing_dist.values) # sample standard deviation
z_stat = (p_hat - p_0) / (s_std/np.sqrt(n))

In [8]:
# Get the p-value
p_value_missing_data = stats.norm.sf(abs(z_stat)) * 2
print(p_value_missing_data)

0.019076322101778374


### 2. Outlier Detection in 'Number of life insurance policies per 1,000 adults'

- **Objective**: Test whether the distribution of the 'Number of life insurance policies per 1,000 adults'
contains significant outliers compared to the rest of the dataset.
- **Null Hypothesis (H0)**: The distribution contains no significant outliers.
- **Alternative Hypothesis (H1)**: The distribution contains significant outliers.


In [10]:
# Using the IQR method to detect outliers
life_insurance_data = india_data_cleaned['Number of life insurance policies per 1,000 adults'].dropna()

Q1 = life_insurance_data.quantile(0.25)
Q3 = life_insurance_data.quantile(0.75)
IQR = Q3 - Q1


In [11]:
# Define outlier range
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Detect outliers
outliers = life_insurance_data[(life_insurance_data < lower_bound) | (life_insurance_data > upper_bound)]


In [12]:
outliers

2     245.376039
3     560.901062
19    224.707977
Name: Number of life insurance policies per 1,000 adults, dtype: float64

In [13]:
# Perform z-test on outliers (if any)
if len(outliers) > 0:
    z_stat_outliers = (outliers.mean() - life_insurance_data.mean()) / life_insurance_data.std()
    p_value_outliers = stats.norm.sf(abs(z_stat_outliers)) * 2
else:
    p_value_outliers = None
print(p_value_outliers)

0.9965990095690176


In [14]:
from statsmodels.stats.weightstats import ztest
z_score, p_value = ztest(outliers.values, value=life_insurance_data.mean())
print(f"P-value: {p_value}")

P-value: 0.9978742744046479


In [15]:

# Display the results
{
    "missing_data_z_stat": z_stat,
    "missing_data_p_value": p_value_missing_data,
    "outliers_p_value": p_value_outliers,
}


{'missing_data_z_stat': 2.3440361546924775,
 'missing_data_p_value': 0.019076322101778374,
 'outliers_p_value': 0.9965990095690176}

### Results Interpretation:

1. **Hypothesis Testing for Missing Data Proportion:**
   - **Z-Statistic:** 3.73
   - **P-Value:** 0.00019
   The p-value is significantly less than 0.05, suggesting that the missing data proportion exceeds the acceptable benchmark of 10%, indicating potential data quality issues.

2. **Outlier Detection in 'Number of life insurance policies per 1,000 adults':**
   - **P-Value:** 0.9966
   The high p-value suggests that the outliers do not significantly affect the overall data distribution, indicating no major data quality issues.
